### PR TITLE
minor: Use md5 for file hashing to reduce possible collisions

### PR DESCRIPTION
**Pull request by mvorisek**
currently collision (with p=50%) is expected with `sqrt(2*2^32*0.5)` [1] = `65536` files only

[1] https://en.wikipedia.org/wiki/Birthday_problem#Square_approximation

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -16,6 +16,7 @@
     "require": {
         "php": "^7.4 || ^8.0",
         "ext-json": "*",
+        "ext-hash": "*",
```
Comment (mvorisek) [inline composer.json:19]: see https://www.php.net/manual/en/hash.installation.php

as of PHP 5.1.2 always enabled when not disabled manually

as of PHP 7.4 always available

File: composer.json [hunk 2]
```diff
@@ -15,6 +15,7 @@
     ],
     "require": {
         "php": "^7.4 || ^8.0",
+        "ext-hash": "*",
```
Comment (mvorisek) [inline composer.json:18]: see https://www.php.net/manual/en/hash.installation.php

as of PHP 5.1.2 always enabled when not disabled manually

as of PHP 7.4 always available

File: tests/Tokenizer/CodeHasherTest.php [hunk 3]
```diff
@@ -26,7 +26,7 @@ final class CodeHasherTest extends TestCase
 {
     public function testCodeHasher(): void
     {
-        static::assertSame('322920910', CodeHasher::calculateCodeHash('<?php echo 1;'));
-        static::assertSame('322920910', CodeHasher::calculateCodeHash('<?php echo 1;')); // calling twice, hashes should always be the same when the input doesn't change.
+        static::assertSame('d9de49676ba2316990a5acd04c8418e8', CodeHasher::calculateCodeHash('<?php echo 1;'));
```
Comment (mvorisek) [inline tests/Tokenizer/CodeHasherTest.php:29]: a collision example is https://3v4l.org/Bc1Da 😃

File: src/Cache/Cache.php [hunk 4]
```diff
@@ -43,7 +43,7 @@ public function has(string $file): bool
         return \array_key_exists($file, $this->hashes);
     }
 
-    public function get(string $file): ?int
+    public function get(string $file): ?string
```
Comment (keradus) [inline src/Cache/Cache.php:46]: after calling `fromJson` with cache file created before your PR, this method will try to return int, not string, due to content in json file

File: tests/Tokenizer/CodeHasherTest.php (hunk 3 above)
Comment (keradus) [reply tests/Tokenizer/CodeHasherTest.php:29]: I wonder how big edge case it is, as nobody complained so far. Did you faced the collision in real codebase? or only managed to generate it?

Comment (mvorisek) [reply tests/Tokenizer/CodeHasherTest.php:29]: As shown in the PR description, with large number of the files, the collision with only 32-hash space is more than expected.

File: src/Cache/Cache.php (hunk 4 above)
Comment (mvorisek) [reply src/Cache/Cache.php:46]: Good catch.

The `signature` will not match as it contains also the PHP CS Fixer `version` in:

https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/v3.10.0/src/Cache/FileCacheManager.php#L113

Comment (keradus) [reply src/Cache/Cache.php:46]: before we call `!$this->signature->equals($cache->getSignature())) {`, we do call
`$cache = $this->handler->read();`
`handler->read()` calls `cache = Cache::fromJson($content);`
`Cache::fromJson()` populates the hashes: https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/v3.10.0/src/Cache/Cache.php#L129 .
so before we would decide to use or not use the Cache data from file due to version check, we already created Cache instance with `hashes` that are `array<string, int>` and not `array<string, string>`, as expected in your branch.

Or I'm missing sth?

Comment (mvorisek) [reply src/Cache/Cache.php:46]: > ... https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/v3.10.0/src/Cache/Cache.php#L129 . so before we would decide to use or not use the Cache data from file due to version check, we already created Cache instance with `hashes` that are `array<string, int>` and not `array<string, string>`, as expected in your branch.

the link is a clear proof the `hashes` from old/crc32 cache will be populated with `array<string, int>`

but my question is, is the cache used somewhere before the signature check? if not, why should be care about this usecase?

or you prefer the `hashes` in https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/v3.10.0/src/Cache/Cache.php#L129 to be always casted to a string?

Comment (keradus) [reply src/Cache/Cache.php:46]: if we claim, and not only with PHP Doc , but also with PHP (strict) type, that given value will be `string`, we should not populate it with anything else.
maybe _now_ we have a flow that given value is not in use, but it's very easy to see that code flow will change. And someone, without touching `get` function here, will start getting it reporting about wrong type on execution time (Eg https://3v4l.org/gje72 ). Let's avoid that.

Yeah, casting or validation can be a solution.

Comment (mvorisek) [reply src/Cache/Cache.php:46]: casting fix landed

**Comment by mvorisek** [comment]
This PR is done. Can it be merged or is there any feedback I should address first?

**Comment by keradus** [comment]
We are looking for community support in _reviewing_ the PRs.
While it's great that the community is giving support in creating new PRs and raising new feature requests / bug reports, it's not realistic for a few ppl to review everything.
Without bigger community support in reviewing the PRs, core members will always be not-as-fast-as-we-would-all-like-to with actions on their side.

File: src/Cache/Cache.php [hunk 5]
```diff
@@ -126,7 +126,11 @@ public static function fromJson(string $json): self
 
         $cache = new self($signature);
 
-        $cache->hashes = $data['hashes'];
+        $cache->hashes = array_map(function ($v): string {
+            // before v3.11.1 the hashes were crc32 encoded and saved as integers
+            // TODO remove the to string cast/array_map in v4.0
```
Comment (keradus) [inline src/Cache/Cache.php:131]:
```suggestion
            // @TODO: remove the to string cast/array_map in v4.0
```

**Pull request merged**